Limit max size in RasterHull()

cwillisf opened the pull request:
`RasterHull()` now limits the size of the `BitmapData` which it creates. The limit is the `texSizeMax` from `DisplayObjectContainerIn3D`, which is now public. Prior to this change the `BitmapData` size was not limited, which could cause an exception to be thrown by the `BitmapData` constructor.

To test:
- Open the browser console
- Load a project which exhibits this problem: https://app.getsentry.com/scratch-team/scratch-flash/issues/105167534/
  (or contact me for an SB2 which triggers the bug)

This is one of at least two issues which cause the symptom described in #1122

Commit:
Limit max size in RasterHull()

`RasterHull()` now limits the size of the `BitmapData` which it creates.
The limit is the `texSizeMax` from `DisplayObjectContainerIn3D`, which
is now public. Prior to this change the `BitmapData` size was not
limited, which could cause an exception to be thrown by the `BitmapData`
constructor.

cwillisf commented on src/scratch/ScratchCostume.as:
```diff
@@ -344,9 +346,16 @@ public class ScratchCostume {
+			desiredHeight *= factor;
+		}
+		// TODO: figure out why we add 1 to each dimension in addition to using Math.ceil above
+		var image:BitmapData = new BitmapData(desiredWidth + 1, desiredHeight + 1, true, 0);
```
Luckily the code below is already set up to use whatever size `image` happens to get here.

cwillisf commented on src/scratch/ScratchCostume.as:
```diff
@@ -344,9 +346,16 @@ public class ScratchCostume {
+				|| desiredHeight >= DisplayObjectContainerIn3D.texSizeMax) {
+			var factor:Number = (DisplayObjectContainerIn3D.texSizeMax - 1) / Math.max(desiredWidth, desiredHeight);
+			desiredWidth *= factor;
+			desiredHeight *= factor;
```
scale while maintaining aspect ratio

thisandagain commented on src/scratch/ScratchCostume.as:
```diff
@@ -344,9 +346,16 @@ public class ScratchCostume {
+		}
+		// TODO: figure out why we add 1 to each dimension in addition to using Math.ceil above
+		var image:BitmapData = new BitmapData(desiredWidth + 1, desiredHeight + 1, true, 0);
 //trace('bitmap rect: '+image.rect);
```
Any idea why this (and some below) commented out code was left in here?

cwillisf replied:
Good point -- I'll remove it.

thisandagain commented on src/render3d/DisplayObjectContainerIn3D.as:
```diff
@@ -26,7 +26,11 @@ package render3d {
 
-public class DisplayObjectContainerIn3D extends Sprite {SCRATCH::allow3d{
+public class DisplayObjectContainerIn3D extends Sprite {
+	public static var texSizeMax:int = 2048;
```
How did you determine these limits?

cwillisf replied:
A while back we allowed up to 4096x4096, but that was causing a lot of noise on Sentry due to systems which didn't support it. The 2048x2048 limit was put in place by @sclements in #897. As he points out in that PR, at that time the top event was "Texture too big (max is 2048x2048)"

There doesn't appear to be any way to query Flash about maximum texture size, but we could increase texture size in a loop until we catch a texture-too-big exception.

Of course, all that is talking about 3D hardware. The change in this PR is important even when 3D acceleration isn't available. I use `texSizeMax` because a) it seems like a reasonable limit and b) it makes sense as a limit when we actually are running in 3D mode.

Commit:
Cleanup around RasterHull()

Use Vector.<Point> instead of untyped Array, as suggested by some
commented code.
Remove the rest of the commented debug code.

TheLogFather commented:
Is it this change which broke this platformer: https://scratch.mit.edu/projects/104297197/ ...?
(See https://scratch.mit.edu/discuss/topic/208322)

cwillisf commented:
@TheLogFather yes, this change probably did cause that project to break. Unfortunately there's no good answer: prior to this change, that project would cause the editor to crash on some computers. Now it doesn't crash on those computers, but it doesn't work right on any computer.

We could adjust the limit according to what each computer is capable of handling, but that would mean projects like this would work on some computers but fail on others. I think that would feel like a bug.

In my opinion it's best to have consistent behavior across as many computers as we reasonably can, and unfortunately right now that means limited support for costumes above 2048 in either dimension.

TheLogFather commented:
Yeah, I suspected as much (hence my recommendation).
